### ISSUE TITLE
TryGetPropertyValue on a nested ComplexType returns T, should return Delta<T>

### ISSUE THREAD

**Issue by dxrdxr** [OData/WebApi#2500]
In an entity that has a nested ComplexType, if the nested resource is changed, it is correctly added to the `_deltaNestedResources` list and all internal state is correct.

Subsequently performing a `TryGetPropertyValue("nested", out var value)` will return the underlying _instance value of type T.

It should return Delta\<T\> if it has changed.

The behavior was added in [this](https://github.com/OData/WebApi/pull/2246) PR to fix another issue.

### Assemblies affected

Latest WebApi (7.5.0? unreleased?).

### Reproduce steps

```
public class Contact 
{
    public int ContactID { get; set; }
    public string Name { get; set; }
    public Address Address { get; set; }
}

public class Address
{
    public int AddressID { get; set; }
    public string Street { get; set; }
}

// Build a nested Delta
var deltaContact = new Delta<Contact>();
var deltaAddress = new Delta<Address>();
deltaContact.TrySetPropertyValue("Name", "Bob Smith");
deltaAddress.TrySetPropertyValue("Street", "123 Main St.");
deltaContact.TrySetPropertyValue("Address", deltaAddress);

// Try of access the nested resource
deltaContact.TryGetPropertyValue("Address", out var address);
Assert.IsTrue(address is not Delta<Address>);
Assert.IsTrue(address is Address);
```

### Expected result

```
Assert.IsTrue(address is Delta<Address>);
Assert.IsTrue(address is Address);
```

### Actual result

```
Assert.IsTrue(address is not Delta<Address>);
Assert.IsTrue(address is Address);
```

### Additional detail
@gathogojr @xuzhg 

The scenario that this breaks is a Controller that receives a Delta for an object that has nested resources. If the Controller simply calls `Patch()` everything works as expected by the internal code recursing on  `_deltaNestedResources`. 

However, if the Controller wants to prevent over-posting of a property in the nested resource, there is no way to determine what has changed as there is no way to get the `Delta<T>` of the nested resource. Without this information the controller must inefficiently copy _all_ properties, defeating the purpose of using `Patch`.

The root cause is [this ](https://github.com/OData/WebApi/blob/f859e032395bee9b4ceec857a34beed36d104d6c/src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs#L673 )code in `TrySetNestedResourceInternal()`.

The code is using Reflection to find the `_instance` of the `deltaNestedResource`. The `_instance` is of type `T`. It then sets the property via:

```
    cacheHit.SetValue(_instance, field.GetValue(deltaNestedResource));
```

But what is actually needed is `Delta<T>`, which conveniently is exactly `deltaNestedResource`, so the code should be:

```
    cacheHit.SetValue(_instance, deltaNestedResource);
```

Edit:
The root cause is actually in the `TryGetPropertyValue` not the Set. The Get should be returning the value from `_deltaNestedResources ` not the `_instance`.

**Comment by habbes** [comment]
@dxrdxr would this be a breaking change?

**Comment by dxrdxr** [comment]
@habbes No. If the user does not set the new property, there is no difference in functionality.

**Comment by dxrdxr** [comment]
@habbes The PR from 8/11/2020 fixed a previous issue where the nested property _instance was not set, the fix was only partial thus introducing this issue.

There is a difference in semantics in what is being returned when a `GetInstance()` is called.

- Before #2246 (7.5.0) the nested property is null
- After #2246 the nested property has the changed value of type `TEntity`
- After this change the nested property will be changed value of type `Delta<TEntity>`

Functionally after the proposed change, the value of the nested property will still contain the changed value, but its type is different.

To get access to the underlying value after this change the caller must either assign it to a `dynamic` variable and access it as if it was type `TEntity`, or to call `TryGetPropertyValue()`.

Is it a "breaking change"? It is only if you consider the first change "breaking" as it changed the semantics, from something unfunctional to functional.

I don't consider it a breaking change as the previous fix was not fully functional and does not fix the real problem, it only fully handles single depth nested properties. 

I wrote a number of test cases with a three levels of properties, each of which has changed and unchanged properties, to work through the details.

Problems with the current implementation:

- There is no possible way to get the Level2 or Level3 Delta<> information
- If a property in Level2 or Level3 has a value that is null (default) there is no way to distinguish between an unchanged property and a property that was changed to null (default). 
- Code that is trying to prevent over-posting cannot use recursively use the common technique of iterating over changed properties and only setting those whose name is known to be "safe".
- The current implementation will properly Patch to arbitrary number of levels, but  only reveals the first level of Delta's.

The proposed change is to put the Delta<T> as the nested property value on the _instance, or when accessing the value through `TryGetPropertyValue()`. The caller must check if it is `typeof(IDelta)` and use it appropriately.

_I can share my test cases if interested_

**Comment by dxrdxr** [comment]
Continuing to frame this issue.

I created a test case that has the following flow (actual test code available on request):

```
public class CustomerEntity
{
    public int ID { get; set; }
    public string FirstName { get; set; }
    public string LastName { get; set; }
    public AddressEntity Address { get; set; }
}
public class AddressEntity
{
    public int ID { get; set; }
    public string StreetAddress { get; set; }
    public string City { get; set; }
    public string State { get; set; }
    public int ZipCode { get; set; }
}
```

Patch with a unchanged Address
- Initialize `deltaAddress = Delta<AddressEntity>` but make no changes
- Initialize `deltaCustomer = Delta<CustomerEntity>` 
- Set `Customer.Address` nested property `deltaCustomer.TrySetPropertyValue("Address", deltaAddress)`
- Get `Customer.Address` nested property `deltaCustomer.TryGetPropertyValue("Address", out object addressProperty)`
- Verify that all of the `Address` properties are default values (i.e. the _instance is uninitialized)
- Patch a known `Customer`
- Verify that the `Customer.Address` has not changed

Patch with a Address.StreetAddress changed to null
- Initialize `deltaAddress = Delta<AddressEntity>` 
- Set Address.StreetAddress property to null `deltaAddress.TrySetPropertyValue("StreetAddress", null)`
- Initialize `deltaCustomer = Delta<CustomerEntity>` 
- Set `Customer.Address` nested property `deltaCustomer.TrySetPropertyValue("Address", deltaAddress)`
- Get `Customer.Address` nested property `deltaCustomer.TryGetPropertyValue("Address", out object addressProperty)`
- Verify that all of the `Address` properties are default values **[This is unexpected!!!]**
- Patch a known `Customer`
- Verify that the `Customer.Address.StreetAddress` has changed to `null`

_Note: in this example I used `TryGetPropertyValue` for simplicity, but the exact same behavior can be observed using `GetInstance`._

The critical problem is that when a Controller receives a `Delta<T>` with a nested property, it cannot distinguish between these two cases: a property of the nested property that is unchanged and a property that is changed to its `default` value. The only way to make the distinction is to actually perform a `Patch` and determine what actually changed.

To re-answer @habbes if this is a breaking change, technically it is breaking. But I will assert that the existing behavior is not useful (broken) because it returns the nested property with properties that cannot accurately be determined if they are changed or not.

Root cause of this may be the `GetInstance` documentation (cited in the previous PR):
> Returns the instance that holds all the changes (and original values) being tracked by this Delta.

The statement "(and original values)" is incorrect, it _never_ returns the original values! It never has access to nor holds the original values.

**NOTE:** This also affects AspNetCoreOData

**Comment by dxrdxr** [comment]
I have come down to two choices to solve this issue:

1. Change the semantics of `TryGetPropertyValue()` to return the actual type that was used in `TrySetPropertyValue`.
2. Add an additional method `TryGetNestedPropertyValue()` that returns the `IDelta<T>` if it was set as an IDelta, otherwise false.

As described above, I greatly prefer # 1 as it provides a cleaner interface, although it is arguably a breaking change. However, the current semantics I believe are broken and not useful as the caller cannot fully determine how to use the nested properties (e.g. if property `A` has a value of `default`, the caller cannot determine if it is an unchanged property or a property changed to be `default`).  

Option # 2 provides proper semantics, but at the expense that it requires the caller to first call the new interface to determine if it is `IDelta`, if false then call the old interface to determine if it has been changed with a value of type T.

I have implemented Option # 1 (along with tests) and have a controller that uses it. To demonstrate how it could be used:

```
//
// Prevent over posting
//
contact.SetUpdatableProperties();
contact.Patch(originalContact);
```
 My Generic extension method on `Delta<T>`:
```
public static void SetUpdatableProperties<T>(this Delta<T> delta) where T : ModelBase<T>
{
    dynamic instance = delta.GetInstance();
    IEnumerable<string> updatableProperties = instance.GetUpdatablePropertyList();

    delta.UpdatableProperties.Clear();
    foreach (var property in updatableProperties)
    {
        delta.UpdatableProperties.Add(property);
    }

    foreach (string nestedPropertyName in instance.GetUpdatableNestedPropertyList())
    {
        if (!delta.TryGetPropertyValue(nestedPropertyName, out dynamic nestedProperty))
        {
            continue;
        }

        if (nestedProperty is IDelta)
        {
            SetUpdatableProperties(nestedProperty);
        }
    }
}
```
Note: For my generic class `ModelBase<T>` each type T implements a method to return a list of updatable properties and nested updatable properties. This abstraction was used to simplify my code. The simple alternative would be to not use generics and just use an explicit type and a hard coded property list.
